Realm/Realm/Sync/Session.cs

```diff
@@ -27,21 +30,76 @@ namespace Realms.Sync
     /// and the server (and a remote Realm served by a MongoDB Realm Server). Sessions are always created by the SDK and vended
     /// out through various APIs. The lifespans of sessions associated with Realms are managed automatically.
     /// </summary>
-    public class Session
+    public class Session : INotifyPropertyChanged
     {
         /// <summary>
         /// Triggered when an error occurs on a session. The <c>sender</c> argument will be the session which has errored.
         /// </summary>
         public static event EventHandler<ErrorEventArgs> Error;
 
+        private readonly SessionHandle _handle;
+
+        [SuppressMessage("StyleCop.CSharp.NamingRules", "SA1300:Element should begin with upper-case letter", Justification = "This is the private event - the public is uppercased.")]
+        private event PropertyChangedEventHandler _propertyChanged;
+
+        private IDisposable _propertyChangedNotificationTokens;
+
+        private SessionHandle Handle
+        {
+            get
+            {
+                if (_handle.IsClosed)
+                {
+                    throw new ObjectDisposedException(
+                        nameof(Session),
+                        "This Session instance is invalid. This typically means that Sync has closed or otherwise invalidated the native session. You can get a new valid instance by calling realm.GetSession().");
+                }
+
+                return _handle;
+            }
+        }
+
         internal bool IsClosed => _handle.IsClosed;
 
+        /// <summary>
+        /// Occurs when a property value changes.
+        /// </summary>
+        public event PropertyChangedEventHandler PropertyChanged
+        {
+            add
+            {
+                if (_propertyChangedNotificationTokens == null)
+                {
+                    SubscribeNotifications();
+                }
+
+                _propertyChanged += value;
+            }
+
+            remove
+            {
+                _propertyChanged -= value;
+
+                if (_propertyChanged == null)
+                {
+                    UnsubscribeNotifications();
+                }
+            }
+        }
+
         /// <summary>
         /// Gets the session’s current state.
         /// </summary>
         /// <value>An enum value indicating the state of the session.</value>
         public SessionState State => Handle.GetState();
 
+        /// <summary>
+        /// Gets the session’s current connection state.
+        /// </summary>
+        /// <remarks>This field notifies clients that its value has changed according to the <see href="https://docs.microsoft.com/en-us/dotnet/api/system.componentmodel.inotifypropertychanged?view=net-6.0">INotifyPropertyChanged</see> interface.</remarks>
```

Comment:
We should remove this remark - the interface itself is implying that this is the case.